internal/reporting.go

```diff
@@ -489,42 +491,177 @@ type ResourceInfo struct {
 	Ingresses   []IngressInfo    `json:"ingresses"`
 }
 
-// FilterableResources returns all of the k8s resources associated with a VICE analysis.
-func (i *Internal) FilterableResources(c echo.Context) error {
-	filter := filterMap(c.Request().URL.Query())
+func (i *Internal) fixUsername(username string) string {
+	if strings.HasSuffix(username, i.UserSuffix) {
+		return username
+	}
+	return fmt.Sprintf("%s%s", username, i.UserSuffix)
+}
 
+func (i *Internal) doResourceListing(filter map[string]string) (*ResourceInfo, error) {
 	deployments, err := i.getFilteredDeployments(filter)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	pods, err := i.getFilteredPods(filter)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	cms, err := i.getFilteredConfigMaps(filter)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	svcs, err := i.getFilteredServices(filter)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	ingresses, err := i.getFilteredIngresses(filter)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	return c.JSON(http.StatusOK, ResourceInfo{
+	return &ResourceInfo{
 		Deployments: deployments,
 		Pods:        pods,
 		ConfigMaps:  cms,
 		Services:    svcs,
 		Ingresses:   ingresses,
-	})
+	}, nil
+}
+
+// AdminDescribeAnalysisHandler returns a listing entry for a single analysis
+// asssociated with the host/subdomain passed in as 'host' from the URL.
+func (i *Internal) AdminDescribeAnalysisHandler(c echo.Context) error {
+	host := c.Param("host")
+
+	// Use the name of the ingress to retrieve the externalID
+	id, err := i.getIDFromHost(host)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusNotFound, err.Error())
+	}
+
+	filter := map[string]string{
+		"subdomain":   host,
+		"external-id": id,
+	}
+
+	listing, err := i.doResourceListing(filter)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+	return c.JSON(http.StatusOK, listing)
+
+}
+
+// DescribeAnalysisHandler returns a listing entry for a single analysis associated
+// with the host/subdomain passed in as 'host' from the URL.
+func (i *Internal) DescribeAnalysisHandler(c echo.Context) error {
+	user := c.QueryParam("user")
+	if user == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, "user query parameter must be set")
+	}
+
+	// Since some usernames don't come through the labelling process unscathed, we have to use
+	// the user ID.
+	fixedUser := i.fixUsername(user)
+	a := apps.NewApps(i.db, i.UserSuffix)
+	_, err := a.GetUserID(fixedUser)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("user %s not found", fixedUser))
+		}
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
```

Comment:
I've seen this error checking code duplicated in a few places, and it doesn't appear that `GetUserID` is called outside of the context of a request handler. I wonder if it might be useful to have `GetUserID` return an `HTTPError` in the case where the username isn't found. Then the callers could return the error unmodified if an error is returned.

Reply:
I wanted to make a function that made no assumptions about what kind of function is calling it. I'd like to keep the HTTP specific stuff out of the apps package, if possible.